### ISSUE TITLE
uv_fs_poll_init design is unsound

### ISSUE THREAD

**Issue by vtjnash** [libuv/libuv#4543]
There is a concurrency mistake in the design of `uv_fs_poll_start`: the documentation promises that the "callback passed to [uv_fs_poll_start()](https://docs.libuv.org/en/v1.x/fs_poll.html#c.uv_fs_poll_cb) which will be called repeatedly after the handle is started, when any change happens to the monitored path."

However, this is not factually accurate to the implementation. The monitoring does not actually start when the uv_fs_poll_start is called, as the documentation claimed, but at some arbitrary time in the distant future when the first uv_poll_cb finally might get a chance to run. A nodejs application that called `uv_fs_poll_init` then immediately called `fs.writeSync` should observe that the file changed (per the libuv and nodejs documentation), but (per the implementation) that stat change will be discarded since there was no recorded state for it to compare against.

In libuv v2, I suggest we delete this uv_fs_poll_t type entirely, and leave it up to individual libraries to build this out of the constituent components (permitting them to choose a design where the user gets some sort of notification with the first stat result / when the first stat result is available to begin polling / initialize their polling with an existing stat result)

**Comment by saghul** [comment]
> A nodejs application that called `uv_fs_poll_init`

I guess you mean start here?



> then immediately called `fs.writeSync` should observe that the file changed (per the libuv and nodejs documentation), but (per the implementation) that stat change will be discarded since there was no recorded state for it to compare against.

I guess we should document thaty non-determinism at least.

> In libuv v2, I suggest we delete this uv_fs_poll_t type entirely, and leave it up to individual libraries to build this out of the constituent components (permitting them to choose a design where the user gets some sort of notification with the first stat result / when the first stat result is available to begin polling / initialize their polling with an existing stat result)

I wouldn't oppose completely scrapping it, since libuv already provides the means to implement it by user applications.

IIRC the idea was to make fsevent use this if the underlying fs didn't support notifications, but it was not possible to detect that reliable, if memory serves right.

**Comment by bnoordhuis** [comment]
> However, this is not factually accurate to the implementation. The monitoring does not actually start when the uv_fs_poll_start is called, as the documentation claimed, but at some arbitrary time in the distant future when the first uv_poll_cb finally might get a chance to run. A nodejs application that called uv_fs_poll_init then immediately called fs.writeSync should observe that the file changed (per the libuv and nodejs documentation), but (per the implementation) that stat change will be discarded since there was no recorded state for it to compare against.

It does work the way the documentation says it does, with a caveat that's a (not straightforward to fix<sup>1</sup>) bug in the implementation.

Here is how it works:

1. uv_fs_poll_init stats the file, saves the result, then
2. starts the poll timer;
3. timer callback starts async  uv_fs_stat
4. stat callback detects changes and invokes the user's uv_fs_poll_cb callback

The bug is that the timer and stat operations are user-observable with `uv_run(loop, UV_RUN_ONCE)` (they're counted as "something happened so drop out of uv_run"), so you have to call it a few times to actually get the uv_fs_poll_cb callback.

Of course most people use UV_RUN_DEFAULT so it's not a problem for them.

<hr>

<sup>1</sup> not straightforward because we'd have to somehow track whether we only processed internal handles and requests

**Comment by vtjnash** [comment]
I don't think that is accurate in my recollection: the first stat happens in the thread pool, so maybe it happens before the first iteration of the loop, but maybe not for a long time either.

**Comment by bnoordhuis** [comment]
I probably misunderstood what you were saying. The first `uv_fs_stat` call is indeed asynchronous (but started from `uv_fs_poll_start`.)

The documentation for `uv_fs_poll_start` says:

> Check the file at `path` for changes every `interval` milliseconds.

Which seems unambiguous to me, but the documentation for `uv_fs_poll_cb` says (what you linked to):

> Callback passed to :c:func:`uv_fs_poll_start` which will be called repeatedly after the handle is started, when any change happens to the monitored path.

Which indeed is somewhat wrong but libuv could be made to work that way by making that first stat call synchronous.

**Comment by vtjnash** [comment]
Sure, it could be synchronous, but blocking there would be also an unexpected API for libuv. There are a lot of breaking changes libuv *could* do for this, but the easiest is to simply delete all of this code and let the application do it correctly at a higher level. We already replaced this with a correctly functioning API in Julia, so I only opened the issue to warn other users of libuv (probably mainly nodejs) that this function is dangerous to rely upon.
